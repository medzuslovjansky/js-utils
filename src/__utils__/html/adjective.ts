/* eslint-disable prettier/prettier */
import type { SteenAdjectiveParadigm } from '../../adjective';
import * as common from './common';

export function render(value: SteenAdjectiveParadigm) {
  return template(common.get(value));
}

export function renderDiff(a: SteenAdjectiveParadigm, b: SteenAdjectiveParadigm) {
  return template(common.diff(a, b));
}

type AdjectiveSubkey =
  | keyof SteenAdjectiveParadigm['singular']
  | keyof SteenAdjectiveParadigm['plural']
  | keyof SteenAdjectiveParadigm['comparison'];

function template(r: (key: keyof SteenAdjectiveParadigm, subkey: AdjectiveSubkey, index?: number) => string | null): string {
  const figure = `\
<figure>
  <figcaption>Adjective declension</figcaption>
  <table>
    <thead>
      <tr>
        <th></th>
        <th colspan="3">singular</th>
      </tr>
      <tr>
        <th>case</th>
        <th>masculine</th>
        <th>neuter</th>
        <th>feminine</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>Nom.</th>
        <td>${r('singular', 'nom', 0) || '—'}</td>
        <td>${r('singular', 'nom', 1) || '—'}</td>
        <td>${r('singular', 'nom', 2) || '—'}</td>
      </tr>
      <tr>
        <th>Acc.</th>
        <td>${r('singular', 'acc', 0) || '—'}</td>
        <td>${r('singular', 'acc', 1) || '—'}</td>
        <td>${r('singular', 'acc', 2) || '—'}</td>
      </tr>
      <tr>
        <th>Gen.</th>
        <td colspan="2">${r('singular', 'gen', 0) || '—'}</td>
        <td>${r('singular', 'gen', 1) || '—'}</td>
      </tr>
      <tr>
        <th>Loc.</th>
        <td colspan="2">${r('singular', 'loc', 0) || '—'}</td>
        <td>${r('singular', 'loc', 1) || '—'}</td>
      </tr>
      <tr>
        <th>Dat.</th>
        <td colspan="2">${r('singular', 'dat', 0) || '—'}</td>
        <td>${r('singular', 'dat', 1) || '—'}</td>
      </tr>
      <tr>
        <th>Ins.</th>
        <td colspan="2">${r('singular', 'ins', 0) || '—'}</td>
        <td>${r('singular', 'ins', 1) || '—'}</td>
      </tr>
    </tbody>
  </table>

  <table>
    <tbody>
      <tr>
        <th></th>
        <th colspan="2">plural</th>
      </tr>
      <tr>
        <th>case</th>
        <th>masculine</th>
        <th>feminine/neuter</th>
      </tr>
      <tr>
        <th>Nom.</th>
        <td>${r('plural', 'nom', 0) || '—'}</td>
        <td>${r('plural', 'nom', 1) || '—'}</td>
      </tr>
      <tr>
        <th>Acc.</th>
        <td>${r('plural', 'acc', 0) || '—'}</td>
        <td>${r('plural', 'acc', 1) || '—'}</td>
      </tr>
      <tr>
        <th>Gen.</th>
        <td colspan="2">${r('plural', 'gen', 0) || '—'}</td>
      </tr>
      <tr>
        <th>Loc.</th>
        <td colspan="2">${r('plural', 'loc', 0) || '—'}</td>
      </tr>
      <tr>
        <th>Dat.</th>
        <td colspan="2">${r('plural', 'dat', 0) || '—'}</td>
      </tr>
      <tr>
        <th>Ins.</th>
        <td colspan="2">${r('plural', 'ins', 0) || '—'}</td>
      </tr>
    </tbody>
  </table>

  <table>
    <tbody>
      <tr>
        <th>Degrees of comparison</th>
        <th>adjective</th>
        <th>adverb</th>
      </tr>
      <tr>
        <th>positive</th>
        <td>${r('comparison', 'positive', 0) || '—'}</td>
        <td>${r('comparison', 'positive', 1) || '—'}</td>
      </tr>
      <tr>
        <th>comparative</th>
        <td>${r('comparison', 'comparative', 0) || '—'}</td>
        <td>${r('comparison', 'comparative', 1) || '—'}</td>
      </tr>
      <tr>
        <th>superlative</th>
        <td>${r('comparison', 'superlative', 0) || '—'}</td>
        <td>${r('comparison', 'superlative', 1) || '—'}</td>
      </tr>
    </tbody>
  </table>
</figure>
`;

  return figure;
}
