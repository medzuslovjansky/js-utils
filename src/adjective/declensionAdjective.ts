/*
 * @link http://steen.free.fr/interslavic/adjectivator.html
 */

import { ALL_CONSONANTS, VOWELS } from '../substitutions';

export function declensionAdjectiveFlat(
  adj: string,
  postfix: string,
  partOfSpeech?: string,
): string[] {
  return getDeclensionAdjectiveFlat(
    declensionAdjective(adj, postfix, partOfSpeech),
  );
}

function getDeclensionAdjectiveFlat(result: any): string[] {
  const forms = [];

  for (const key of Object.keys(result)) {
    const value = result[key];
    if (typeof value === 'string') {
      forms.push(value);
    } else if (value !== undefined) {
      const notFlat: any = Object.values(value);
      const flatArr = notFlat
        .flat()
        .map((word: string) => word.replace(/ /g, '').split('/'))
        .flat();
      forms.push(...flatArr);
    }
  }

  return Array.from(new Set(forms.filter(Boolean)));
}

/** @deprecated */
export type SteenAdjectiveParadigm = {
  singular: SteenAdjectiveParadigm$Case;
  plural: SteenAdjectiveParadigm$Case;
  short: string | undefined;
  comparison: SteenAdjectiveParadigm$Comparison;
};

type SteenAdjectiveParadigm$Case = {
  nom: string[];
  acc: string[];
  gen: string[];
  loc: string[];
  dat: string[];
  ins: string[];
};

type SteenAdjectiveParadigm$Comparison = {
  positive: string[];
  comparative: string[];
  superlative: string[];
};

const LOOKS_LIKE_ADJECTIVE = /[yij]$/;

function looksLikeAdjective(word: string) {
  return LOOKS_LIKE_ADJECTIVE.test(word);
}

export function declensionAdjective(
  adj: string,
  postfix = '',
  partOfSpeech = 'adj.',
): SteenAdjectiveParadigm {
  const isCompound = adj.includes(' ');
  if (isCompound) {
    const words = adj.split(' ');
    const index = words.findIndex(looksLikeAdjective);
    if (0 <= index && index < words.length - 1) {
      postfix = ' ' + words.slice(index + 1).join(' ') + postfix;
      adj = words.slice(0, index + 1).join(' ');
    }
  }

  const root = establish_root(adj);
  const short = short_form(adj, root, partOfSpeech);
  const m_nom_sg = m_nominative_sg(adj, root);
  const m_acc_sg = m_accusative_sg(adj, root);
  const f_nom_sg = f_nominative_sg(root);
  const n_nom_sg = n_nominative_sg(root);
  const mn_gen_sg = mn_genitive_sg(root);
  const mn_dat_sg = mn_dative_sg(root);
  const mn_ins_sg = mn_instrumental_sg(root);
  const mn_loc_sg = mn_locative_sg(root);
  const f_acc_sg = f_accusative_sg(root);
  const f_gdl_sg = f_gendatloc_sg(root);
  const f_ins_sg = f_instrumental_sg(root);

  const m_nom_pl = m_nominative_pl(root);
  const m_acc_pl = m_accusative_pl(root);
  const fn_nom_pl = fn_nominative_pl(root);
  const glo_pl = genloc_pl(root);
  const dat_pl = dative_pl(root);
  const ins_pl = instrumental_pl(root);

  return {
    singular: {
      nom: applyRules([m_nom_sg, n_nom_sg, f_nom_sg], postfix),
      acc: applyRules([m_acc_sg, n_nom_sg, f_acc_sg], postfix),
      gen: applyRules([mn_gen_sg, f_gdl_sg], postfix),
      loc: applyRules([mn_loc_sg, f_gdl_sg], postfix),
      dat: applyRules([mn_dat_sg, f_gdl_sg], postfix),
      ins: applyRules([mn_ins_sg, f_ins_sg], postfix),
    },
    plural: {
      nom: applyRules([m_nom_pl, fn_nom_pl], postfix),
      acc: applyRules([m_acc_pl, fn_nom_pl], postfix),
      gen: applyRules([glo_pl], postfix),
      loc: applyRules([glo_pl], postfix),
      dat: applyRules([dat_pl], postfix),
      ins: applyRules([ins_pl], postfix),
    },
    short,
    comparison: isCompound
      ? { positive: [], comparative: [], superlative: [] }
      : deriveComparison(root, adj, postfix, partOfSpeech),
  };
}

function deriveComparison(
  root: string,
  adj: string,
  postfix: string,
  partOfSpeech: string,
): SteenAdjectiveParadigm['comparison'] {
  const { isPositive, isComparative, isSuperlative } =
    parseComparisionModifiers(partOfSpeech);

  const adv = adverb(root);
  const m_nom_sg = m_nominative_sg(adj, root);
  const comp_adj = isComparative ? m_nom_sg : comparative_adj(root);
  const comp_adv = isComparative ? adv : comparative_adv(root);
  const sup_adj = isSuperlative ? m_nom_sg : superlative(root, comp_adj, 'adj');
  const sup_adv = isSuperlative ? adv : superlative(root, comp_adv, 'adv');

  return {
    positive: isPositive ? applyRules([m_nom_sg, adv], postfix) : [],
    comparative: isSuperlative ? [] : applyRules([comp_adj, comp_adv], postfix),
    superlative: applyRules([sup_adj, sup_adv], postfix),
  };
}

function parseComparisionModifiers(partOfSpeech: string) {
  const isComparative = partOfSpeech.endsWith('comp.');
  const isSuperlative = partOfSpeech.endsWith('sup.');

  return {
    isPositive: !isComparative && !isSuperlative,
    isComparative,
    isSuperlative,
  };
}

function applyRules(arr: string[], postfix: string) {
  return arr.map(rules).map((item) => {
    return item.replace(/$/, postfix).replace(/( *\/ *)/g, postfix + '$1');
  });
}

const ROOT_ESTABLISHER_RULES: [RegExp, string][] = [
  [/^(?:s[eė]j|sjej)$/, 's|^'],
  [/^v[eė][sś]$/, 'vs|^'],
  [/^(o[vn])(?:oj)?$/, '$1|'],
  [/^(.*)jedin$/, '$1jedn|'],
  [/^(.*)[ėȯ]j$/, '$1|'],
  [/^(.*)[ėȯ](.|[ln]j)$/, '$1$2|'],
  [/^(.*)en$/, '$1n|'],
  [/^.*(?:rad|in|[oe]v)$/, '$&|'],
  [/^(.*t)[oȯ]j$/, '$1|'],
  [/^(?:naš|vaš|.*čij|.*oj)$/, '$&|^'],
  [/^(.*)(?:ij?)$/, '$1^'],
  [/^(.*)(?:yj?)$/, '$1'],
];

const KGH_CARET = /[kgh]\^/g;

function establish_root(adj: string) {
  for (const [pattern, replacement] of ROOT_ESTABLISHER_RULES) {
    if (pattern.test(adj)) {
      const result = adj.replace(pattern, replacement);
      return result.replace(KGH_CARET, (match) => match[0]);
    }
  }

  return '';
}

// TODO: should these words have PoS like "pron.indef." instead of just "adj."?
const NO_SHORT_FORM = /(^in|ktor|seksi|sk|[čj]\^|\|)$/;
const NJI_ENDING = /[^aåeęěijrŕouųy]nj\^$/;

const SIMPLEST_FORM_EXCEPTIONS = /(sinj)$/;
const SHORT_O_ENDING = /(z[gk]|[mz]l|[^jcćčšžŕĺľťśď]k)$/;
const SHORT_E_ENDING = /(bl|[^l][kn])$/;
const MAYBE_COMPARATIVE = /[ćš]\^$/;
const TWO_LETTER_ENDING = /^.[^č]$/; // "věth/trězv/dobr", but not "rabotč/izkonavč"
const SYLLABIC_ENDING_REGEX = new RegExp(
  `([${VOWELS}]st[ŕrl]|tv[rŕ]d|lst|č[rŕ]stv|m[rŕ]tv|brz)$`,
); // ostr, bystr, črstv, tvrd, mrtv, brz
const CONSONANTS_ENDING = new RegExp(`[j${ALL_CONSONANTS}]*$`);

function short_form(
  adj: string,
  root: string,
  partOfSpeech: string,
): string | undefined {
  if (!adj.endsWith('y') && !adj.endsWith('i')) {
    return;
  }

  const simplestForm = adj.slice(0, -1);
  if (SIMPLEST_FORM_EXCEPTIONS.test(root)) {
    return simplestForm; // simplestForm;
  }

  if (NJI_ENDING.test(root)) {
    root = root.slice(0, -2);
  }

  if (NO_SHORT_FORM.test(root)) {
    return;
  }

  if (MAYBE_COMPARATIVE.test(root)) {
    const modifiers = parseComparisionModifiers(partOfSpeech);
    return modifiers.isComparative || modifiers.isSuperlative
      ? undefined
      : simplestForm;
  }

  if (root.endsWith('^')) {
    return simplestForm;
  }

  // should we add yers?
  const [consEnding] = root.match(CONSONANTS_ENDING)!;
  if (consEnding.length === 1) {
    return root;
  }

  if (SHORT_O_ENDING.test(root)) {
    return root.slice(0, -1) + 'ȯ' + root.slice(-1); // zl -> zȯl
  }
  if (SHORT_E_ENDING.test(root)) {
    return root.slice(0, -1) + 'ė' + root.slice(-1); // mekky -> mekek
  }
  if (TWO_LETTER_ENDING.test(consEnding) || SYLLABIC_ENDING_REGEX.test(root)) {
    return root;
  }

  return;
}

function m_nominative_sg(adj: string, root: string) {
  if (root.indexOf('|') !== -1) {
    return adj;
  } else {
    return root + 'y';
  }
}

function f_nominative_sg(root: string) {
  return root + 'a';
}

function n_nominative_sg(root: string) {
  return root + 'o';
}

function mn_genitive_sg(root: string) {
  return root + 'ogo';
}

function mn_dative_sg(root: string) {
  return root + 'omu';
}

function m_accusative_sg(adj: string, root: string) {
  return root + 'ogo / ' + adj;
}

function mn_instrumental_sg(root: string) {
  return root + 'ym';
}

function mn_locative_sg(root: string) {
  return root + 'om';
}

function f_accusative_sg(root: string) {
  return root + 'ų';
}

function f_gendatloc_sg(root: string) {
  return root + 'oj';
}

function f_instrumental_sg(root: string) {
  return root + 'ojų';
}

function m_nominative_pl(root: string) {
  return root + 'i / ' + root + 'e';
}

function m_accusative_pl(root: string) {
  return root + 'yh / ' + root + 'e';
}

function fn_nominative_pl(root: string) {
  return root + 'e';
}

function genloc_pl(root: string) {
  return root + 'yh';
}

function dative_pl(root: string) {
  return root + 'ym';
}

function instrumental_pl(root: string) {
  return root + 'ymi';
}

function adverb(root: string) {
  if (root.charAt(root.length - 2) == 'ć') {
    return root + 'i';
  } else {
    return root + 'o';
  }
}

function comparative_adj(root: string) {
  let result = '';
  const hacek = root.indexOf('^');
  const lastchar = hacek !== -1 ? root.length - 2 : root.length - 1;
  const vowel = /[aåeěęėioȯuųy]/;
  // const liquid = /[lrŕ]/;
  // const nasal = /[nm]/;

  if (root == 'velik') {
    result = 'vęčši';
  } else if (root == 'mal') {
    result = 'menši';
  } else if (root == 'dobr') {
    result = 'lěpši, lučši';
  } else if (root == 'zl') {
    result = 'gorši';
  } else if (root == 'mnog') {
    result = 'boľši';
  } else if (root == 'blåg' || root == 'blag') {
    result = 'unši, ' + root.substring(0, root.length - 1) + 'žejši';
  } else if (root.lastIndexOf('sk') == root.length - 2) {
    result = 'bolje ' + root + 'i';
  } else if (
    root.lastIndexOf('ok') == root.length - 2 ||
    root.lastIndexOf('ek') == root.length - 2
  ) {
    result = root.substring(0, root.length - 2) + 'ši';
  } else if (
    root.lastIndexOf('k') == root.length - 1 &&
    vowel.test(root.charAt(lastchar - 1)) == false
  ) {
    result = root.substring(0, root.length - 1) + 'ši';
  } else {
    if (hacek == -1) {
      result = root + '%ějši';
    } else {
      result = root + '%ejši';
    }
  }
  result = result
    .replace(/k%/, 'č')
    .replace(/g%/, 'ž')
    .replace(/h%/, 'š')
    .replace(/lši/, 'ľši')
    .replace(/gši/, 'žši')
    .replace(/ležši/, 'legši')
    .replace(/%/, '');

  return result;
}

function comparative_adv(root: string) {
  let result = '';
  const hacek = root.indexOf('^');
  const lastchar = hacek !== -1 ? root.length - 2 : root.length - 1;
  const vowel = /[aåeěęėioȯuųy]/;
  // const liquid = /[lrŕ]/;

  if (root == 'velik') {
    result = 'vęče';
  } else if (root == 'mal') {
    result = 'menje';
  } else if (root == 'dobr') {
    result = 'lěpje, lučše';
  } else if (root == 'zl') {
    result = 'gorje';
  } else if (root == 'mnog') {
    result = 'bolje';
  } else if (root == 'blåg' || root == 'blag') {
    result = 'unje, ' + root.substring(0, root.length - 1) + 'žeje';
  } else if (root.lastIndexOf('sk') == root.length - 2) {
    result = 'bolje ' + root + 'o';
  } else if (
    root.lastIndexOf('ok') == root.length - 2 ||
    root.lastIndexOf('ek') == root.length - 2
  ) {
    result = root.substring(0, root.length - 2) + '%je';
  } else if (
    root.lastIndexOf('k') == root.length - 1 &&
    vowel.test(root.charAt(lastchar - 1)) == false
  ) {
    result = root.substring(0, root.length - 1) + '%je';
  } else if (root.indexOf('^') != -1) {
    result = root + 'eje';
  } else {
    result = root + '%ěje';
  }
  result = result
    .replace(/k%ě/, 'če')
    .replace(/g%ě/, 'že')
    .replace(/h%ě/, 'še')
    .replace(/k%j/, 'kš')
    .replace(/g%j/, 'gš')
    .replace(/st%j/, 'šč')
    .replace(/s%j/, 'š')
    .replace(/z%j/, 'ž')
    .replace(/t%j/, 'č')
    .replace(/d%j/, 'dž')
    .replace(/%/, '');

  return result;
}

function superlative(root: string, comp: unknown, srt: string) {
  if (root === 'dobr' && srt === 'adj') {
    return 'najlěpši, najlučši';
  }
  if (root === 'dobr' && srt === 'adv') {
    return 'najlěpje, najlučše';
  }
  if (root === 'blåg' && srt === 'adj') {
    return 'najunši, najblåžejši';
  }
  if (root === 'blag' && srt === 'adj') {
    return 'najunši, najblažejši';
  }
  if (root === 'blåg' && srt === 'adv') {
    return 'najunje, najblåžeje';
  }
  if (root === 'blag' && srt === 'adv') {
    return 'najunje, najblažeje';
  }

  return 'naj' + comp;
}

function rules(word: string): string {
  return (
    word
      .replace(/\^o/g, '^e')
      .replace(/\^y/g, '^i')
      .replace(/s\|\^e/g, 'se')
      .replace(/s\|\^i/g, 'si')
      .replace(/\|/g, '')
      /*.replace('č^', 'č')
        .replace('š^', 'š')
        .replace('ž^', 'ž')
        .replace('ć^', 'ć')
        .replace('c^', 'c')*/
      .replace(/l\^/g, 'lj')
      .replace(/n\^/g, 'ń')
      .replace(/r\^/g, 'ŕ')
      .replace(/j\^/g, 'j')
      .replace(/t\^/g, 'ť')
      .replace(/d\^/g, 'ď')
      .replace(/s\^/g, 'ś')
      .replace(/z\^/g, 'ź')
      .replace(/\^/g, '')
      .replace(/jy/g, 'ji')
      .replace(/cy/g, 'ci')
  );
}
