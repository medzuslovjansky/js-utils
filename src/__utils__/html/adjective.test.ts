/* eslint-disable @typescript-eslint/no-non-null-assertion */

import * as adjective from './adjective';
import { declensionAdjective } from '../../adjective';

describe('renderers', () => {
  describe('HTML', () => {
    describe('noun declension', () => {
      test('render', () => {
        // This word does not exist, but it demonstrates the rendering
        // In alternative reality, it would mean "the best of all ever" :)
        const najdobry_koli = declensionAdjective(
          'najdobry',
          '-koli',
          'adj.sup.',
        )!;
        expect(adjective.render(najdobry_koli)).toMatchSnapshot();
      });

      test('renderDiff', () => {
        const before = declensionAdjective('svěži');
        // we choose a different noun to show the diff
        const after = declensionAdjective('črstvy');
        expect(adjective.renderDiff(before, after)).toMatchSnapshot();
      });
    });
  });
});
