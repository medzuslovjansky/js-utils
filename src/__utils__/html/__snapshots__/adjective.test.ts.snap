// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderers HTML noun declension render 1`] = `
"<figure>
  <figcaption>Adjective declension</figcaption>
  <table>
    <thead>
      <tr>
        <th></th>
        <th colspan="3">singular</th>
      </tr>
      <tr>
        <th>case</th>
        <th>masculine</th>
        <th>neuter</th>
        <th>feminine</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>Nom.</th>
        <td>najdobry-koli</td>
        <td>najdobro-koli</td>
        <td>najdobra-koli</td>
      </tr>
      <tr>
        <th>Acc.</th>
        <td>najdobrogo-koli / najdobry-koli</td>
        <td>najdobro-koli</td>
        <td>najdobrų-koli</td>
      </tr>
      <tr>
        <th>Gen.</th>
        <td colspan="2">najdobrogo-koli</td>
        <td>najdobroj-koli</td>
      </tr>
      <tr>
        <th>Loc.</th>
        <td colspan="2">najdobrom-koli</td>
        <td>najdobroj-koli</td>
      </tr>
      <tr>
        <th>Dat.</th>
        <td colspan="2">najdobromu-koli</td>
        <td>najdobroj-koli</td>
      </tr>
      <tr>
        <th>Ins.</th>
        <td colspan="2">najdobrym-koli</td>
        <td>najdobrojų-koli</td>
      </tr>
    </tbody>
  </table>

  <table>
    <tbody>
      <tr>
        <th></th>
        <th colspan="2">plural</th>
      </tr>
      <tr>
        <th>case</th>
        <th>masculine</th>
        <th>feminine/neuter</th>
      </tr>
      <tr>
        <th>Nom.</th>
        <td>najdobri-koli / najdobre-koli</td>
        <td>najdobre-koli</td>
      </tr>
      <tr>
        <th>Acc.</th>
        <td>najdobryh-koli / najdobre-koli</td>
        <td>najdobre-koli</td>
      </tr>
      <tr>
        <th>Gen.</th>
        <td colspan="2">najdobryh-koli</td>
      </tr>
      <tr>
        <th>Loc.</th>
        <td colspan="2">najdobryh-koli</td>
      </tr>
      <tr>
        <th>Dat.</th>
        <td colspan="2">najdobrym-koli</td>
      </tr>
      <tr>
        <th>Ins.</th>
        <td colspan="2">najdobrymi-koli</td>
      </tr>
    </tbody>
  </table>

  <table>
    <tbody>
      <tr>
        <th>Degrees of comparison</th>
        <th>adjective</th>
        <th>adverb</th>
      </tr>
      <tr>
        <th>positive</th>
        <td>—</td>
        <td>—</td>
      </tr>
      <tr>
        <th>comparative</th>
        <td>—</td>
        <td>—</td>
      </tr>
      <tr>
        <th>superlative</th>
        <td>najdobry-koli</td>
        <td>najdobro-koli</td>
      </tr>
    </tbody>
  </table>
</figure>
"
`;

exports[`renderers HTML noun declension renderDiff 1`] = `
"<figure>
  <figcaption>Adjective declension</figcaption>
  <table>
    <thead>
      <tr>
        <th></th>
        <th colspan="3">singular</th>
      </tr>
      <tr>
        <th>case</th>
        <th>masculine</th>
        <th>neuter</th>
        <th>feminine</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>Nom.</th>
        <td><del>svěži</del> <ins>črstvy</ins></td>
        <td><del>svěže</del> <ins>črstvo</ins></td>
        <td><del>svěža</del> <ins>črstva</ins></td>
      </tr>
      <tr>
        <th>Acc.</th>
        <td><del>svěžego / svěži</del> <ins>črstvogo / črstvy</ins></td>
        <td><del>svěže</del> <ins>črstvo</ins></td>
        <td><del>svěžų</del> <ins>črstvų</ins></td>
      </tr>
      <tr>
        <th>Gen.</th>
        <td colspan="2"><del>svěžego</del> <ins>črstvogo</ins></td>
        <td><del>svěžej</del> <ins>črstvoj</ins></td>
      </tr>
      <tr>
        <th>Loc.</th>
        <td colspan="2"><del>svěžem</del> <ins>črstvom</ins></td>
        <td><del>svěžej</del> <ins>črstvoj</ins></td>
      </tr>
      <tr>
        <th>Dat.</th>
        <td colspan="2"><del>svěžemu</del> <ins>črstvomu</ins></td>
        <td><del>svěžej</del> <ins>črstvoj</ins></td>
      </tr>
      <tr>
        <th>Ins.</th>
        <td colspan="2"><del>svěžim</del> <ins>črstvym</ins></td>
        <td><del>svěžejų</del> <ins>črstvojų</ins></td>
      </tr>
    </tbody>
  </table>

  <table>
    <tbody>
      <tr>
        <th></th>
        <th colspan="2">plural</th>
      </tr>
      <tr>
        <th>case</th>
        <th>masculine</th>
        <th>feminine/neuter</th>
      </tr>
      <tr>
        <th>Nom.</th>
        <td><del>svěži / svěže</del> <ins>črstvi / črstve</ins></td>
        <td><del>svěže</del> <ins>črstve</ins></td>
      </tr>
      <tr>
        <th>Acc.</th>
        <td><del>svěžih / svěže</del> <ins>črstvyh / črstve</ins></td>
        <td><del>svěže</del> <ins>črstve</ins></td>
      </tr>
      <tr>
        <th>Gen.</th>
        <td colspan="2"><del>svěžih</del> <ins>črstvyh</ins></td>
      </tr>
      <tr>
        <th>Loc.</th>
        <td colspan="2"><del>svěžih</del> <ins>črstvyh</ins></td>
      </tr>
      <tr>
        <th>Dat.</th>
        <td colspan="2"><del>svěžim</del> <ins>črstvym</ins></td>
      </tr>
      <tr>
        <th>Ins.</th>
        <td colspan="2"><del>svěžimi</del> <ins>črstvymi</ins></td>
      </tr>
    </tbody>
  </table>

  <table>
    <tbody>
      <tr>
        <th>Degrees of comparison</th>
        <th>adjective</th>
        <th>adverb</th>
      </tr>
      <tr>
        <th>positive</th>
        <td><del>svěži</del> <ins>črstvy</ins></td>
        <td><del>svěže</del> <ins>črstvo</ins></td>
      </tr>
      <tr>
        <th>comparative</th>
        <td><del>svěžejši</del> <ins>črstvějši</ins></td>
        <td><del>svěžeje</del> <ins>črstvěje</ins></td>
      </tr>
      <tr>
        <th>superlative</th>
        <td><del>najsvěžejši</del> <ins>najčrstvějši</ins></td>
        <td><del>najsvěžeje</del> <ins>najčrstvěje</ins></td>
      </tr>
    </tbody>
  </table>
</figure>
"
`;
